haystack/nodes/retriever/_embedding_encoder.py

```diff
@@ -374,9 +377,71 @@ def save(self, save_dir: Union[Path, str]):
         )
 
 
+class _OpenAIEmbeddingEncoder(_BaseEmbeddingEncoder):
+    def __init__(self, retriever: "EmbeddingRetriever"):
+        # pretrained embedding models coming from:
+        self.max_seq_len = retriever.max_seq_len
+        self.url = "https://api.openai.com/v1/embeddings"
+        self.api_key = retriever.api_key
+        model_class: str = next(
+            (m for m in ["ada", "babbage", "davinci", "curie"] if m in retriever.embedding_model), "babbage"
+        )
+        self.query_model_encoder_engine = f"text-search-{model_class}-query-001"
+        self.doc_model_encoder_engine = f"text-search-{model_class}-doc-001"
+        self.tokenizer = AutoTokenizer.from_pretrained("gpt2")
+
+    def ensure_texts_limit(self, text: str):
+        tokenized_payload = self.tokenizer(text)
+        return self.tokenizer.decode(tokenized_payload["input_ids"][: self.max_seq_len])
+
+    def embed(self, model, text: str) -> np.ndarray:
+        payload = {"model": model, "input": self.ensure_texts_limit(text)}
+        headers = {"Authorization": f"Bearer {self.api_key}", "Content-Type": "application/json"}
+        response = requests.request("POST", self.url, headers=headers, data=json.dumps(payload), timeout=30)
+        res = json.loads(response.text)
+
+        if response.status_code != 200:
+            raise OpenAIError(
+                f"OpenAI returned an error.\n"
+                f"Status code: {response.status_code}\n"
+                f"Response body: {response.text}"
+            )
+
+        generated_embeddings = [ans["embedding"] for ans in res["data"]]
+        return np.array(generated_embeddings)
+
+    def embed_queries(self, queries: List[str]) -> np.ndarray:
+        embeddings: List[np.ndarray] = []
+        for query in queries:
+            embedding = self.embed(self.query_model_encoder_engine, query)
+            embeddings.append(embedding)
+        return np.concatenate(embeddings)
+
+    def embed_documents(self, docs: List[Document]) -> np.ndarray:
+        embeddings: List[np.ndarray] = []
+        for doc in docs:
+            embedding = self.embed(self.doc_model_encoder_engine, doc.content)
+            embeddings.append(embedding)
```

Comment:
According to [OpenAI Documentation](https://beta.openai.com/docs/api-reference/embeddings/create), we can get embeddings for multiple inputs in a single request. My guess is that this would probably be a bit more efficient than doing one request for each Document.

Also, we should probably take care of [OpenAI's rate limit](https://help.openai.com/en/articles/5955598-is-api-usage-subject-to-any-rate-limits) given that we usually create embeddings for a large number of Documents. Timo worked in #3078 on a solution for the `OpenAIAnswerGenerator` (this PR got unfortunately stale). Other than that, you might also want to take a look at [this notebook by OpenAI](https://github.com/openai/openai-cookbook/blob/main/examples/How_to_handle_rate_limits.ipynb) on best practices for rate limit handling.